transport/src/test/java/io/netty/channel/DefaultChannelPipelineTailTest.java

```diff
@@ -0,0 +1,402 @@
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import io.netty.bootstrap.Bootstrap;
+
+public class DefaultChannelPipelineTailTest {
+
+    private static EventLoopGroup GROUP;
+
+    @BeforeClass
+    public static void init() {
+        GROUP = new DefaultEventLoopGroup(1);
+    }
+
+    @AfterClass
+    public static void destroy() {
+        GROUP.shutdownGracefully();
+    }
+
+    @Test
+    public void testOnUnhandledInboundChannelActive() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(1);
+        MyChannel myChannel = new MyChannel() {
+            @Override
+            protected void onUnhandledInboundChannelActive() {
+                latch.countDown();
+            }
+        };
+
+        Bootstrap bootstrap = new Bootstrap()
+                .channelFactory(new MyChannelFactory(myChannel))
+                .group(GROUP)
+                .handler(new ChannelInboundHandlerAdapter())
+                .remoteAddress(new InetSocketAddress(0));
+
+        Channel channel = bootstrap.connect()
+                .syncUninterruptibly().channel();
+
+        try {
+            assertTrue(latch.await(1L, TimeUnit.SECONDS));
+        } finally {
+            channel.close();
+        }
+    }
+
+    @Test
+    public void testOnUnhandledInboundChannelInactive() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(1);
+        MyChannel myChannel = new MyChannel() {
+            @Override
+            protected void onUnhandledInboundChannelInactive() {
+                latch.countDown();
+            }
+        };
+
+        Bootstrap bootstrap = new Bootstrap()
+                .channelFactory(new MyChannelFactory(myChannel))
+                .group(GROUP)
+                .handler(new ChannelInboundHandlerAdapter())
+                .remoteAddress(new InetSocketAddress(0));
+
+        Channel channel = bootstrap.connect()
+                .syncUninterruptibly().channel();
+
+        channel.close().syncUninterruptibly();
+
+        assertTrue(latch.await(1L, TimeUnit.SECONDS));
+    }
+
+    @Test
+    public void testOnUnhandledInboundException() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(1);
+        MyChannel myChannel = new MyChannel() {
+            @Override
+            protected void onUnhandledInboundException(Throwable cause) {
+                latch.countDown();
+            }
+        };
+
+        Bootstrap bootstrap = new Bootstrap()
+                .channelFactory(new MyChannelFactory(myChannel))
+                .group(GROUP)
+                .handler(new ChannelInboundHandlerAdapter())
+                .remoteAddress(new InetSocketAddress(0));
+
+        Channel channel = bootstrap.connect()
+                .syncUninterruptibly().channel();
+
+        try {
+            channel.pipeline().fireExceptionCaught(new IOException("testOnUnhandledInboundException"));
+            assertTrue(latch.await(1L, TimeUnit.SECONDS));
```

Comment:
you might want to check that you got the right exception here.   Any exception, perhaps one you weren't expecting, would make this test pass.